performance/index.ts

```diff
@@ -0,0 +1,156 @@
+import { EMPTY, from, Observable, Subscription } from 'rxjs';
+import { tap } from 'rxjs/operators';
+
+type FirebaseApp = import('firebase/app').FirebaseApp;
+
+/**
+ * Lazy loads Firebase Performance monitoring and returns the instance as 
+ * an observable
+ * @param app 
+ * @returns Observable<FirebasePerformance>
+ */
+export const getPerformance$ = (app: FirebaseApp) => from(
+  import('firebase/performance').then(module => module.getPerformance(app))
+);
+
+/**
+ * Creates an observable that begins a trace with a given id. The trace is ended
+ * when the observable unsubscribes. The measurement is also logged as a performance
+ * entry.
+ * @param traceId 
+ * @returns Observable<void>
+ */
+const trace$ = (traceId: string) => {
+  if (typeof window !== 'undefined' && window.performance) {
+    const entries = window.performance.getEntriesByName(traceId, 'measure') || [];
+    const startMarkName = `_${traceId}Start[${entries.length}]`;
+    const endMarkName = `_${traceId}End[${entries.length}]`;
+    return new Observable<void>(emitter => {
+      window.performance.mark(startMarkName);
+      emitter.next();
+      return {
+        unsubscribe: () => {
+          window.performance.mark(endMarkName);
+          window.performance.measure(traceId, startMarkName, endMarkName);
+        }
+      };
+    });
+  } else {
+    return EMPTY;
+  }
+};
+
+/**
+ * Creates a function that creates an observable that begins a trace with a given id. The trace is ended
+ * when the observable unsubscribes. The measurement is also logged as a performance
+ * entry.
+ * @param name 
+ * @returns (source$: Observable<T>) => Observable<T>
+ */
+export const trace = <T = any>(name: string) => (source$: Observable<T>) => new Observable<T>(subscriber => {
```

Comment:
Should these `trace*` functions do some kind of check that Firebase Performance exists? Since we're (rightfully) using the browser's built in `window.performance`, I'm worried that someone might call `trace` without any errors or warnings, but can't figure out why their traces aren't showing up in the Firebase console

Reply:
We talked about this a bit in our meeting last week. I think the conclusion was that keeping perf out of the way was intentional, but in the docs we'll need to remind developers to import it at _some_ point so that the data makes it to perf.